--
title: "Coding, Non-coding, and SpliceAI"
date: "`r Sys.time()`"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    css: ~/R/bootstrap.css
    #css: bootstrap.css
    fig_width: 12 
    fig_height: 12

editor_options:
  chunk_output_type: inline
---
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

# Loading packages
```{r setup, include=FALSE}
library(tidyverse)
library(ggridges)
library(ggthemes)
library(patchwork)
library(caret)
library(e1071)
library(ggpubr)
library(GGally)
library(plotly)
try(
  source('~/R/functions.R')  
)
try(source('functions.R'))
```


# Define mutable data
```{r}
Data = NULL
SO_LEVELS = c("Missense", "Synonymous", "StopGain")









vec = gsub(SAMPLE_PREFIX, "", LEVELS)
Nu_Sep = nchar(SAMPLE_PREFIX)
Contr_rat <-  paste(vec[-1], "_cont_rat", sep = "")
Optimal_Date <- Contr_rat[length(Contr_rat)]
OUTDIR = data_path
```

# Read in metrics

Table Description

  * `sample_name`:  sample_name,
  * `total_count`:  total_count, Total number of reads in the library
  * `off_target`: Reads that didn't start in the right posiiton
  * `cigar_count`:  Fails CIGAR filter (150 contiguous bases)
  * `mm_count`:     Has too many amino acid changes
  * `invalid_count`: Required DNA changes not present
  * `frameshift_count`:  How many frameshifts
  * `nochange_count`: How many with no changes
  * `synonymous_count`: How many synonymous?
  * `missense_count`: How many missense
  * `stop_count`: How many stop-gained mutations
  * `mixed_count`: How many Synonmymous-Missense hybrid reads
  * `multi_syn`: How many reads with multiple synonymous variants were found
  * `usable_count`:  synonymous_count + missense_count + stop_count


```{r eval=TRUE}
METRICS = NULL
for (x in list.files(data_path, pattern = 'metrics.tsv', full.names = TRUE)) {
  tmp = read.csv(x, sep = "|", stringsAsFactors = FALSE)
  tmp = tmp[-1,]
  METRICS = rbind(tmp, METRICS)
}
METRICS = METRICS %>%
  select(-X, -X.1) %>%
  separate(sample_name, into = "Sample", extra = "drop", sep = '\\.') %>%
  mutate(Sample = gsub("_hy1","", Sample)) %>%
  mutate(Sample = gsub("_hyl","", Sample)) %>%
  mutate(usable_percent = usable_count) %>%
  separate(col = "usable_count", into="usable", extra = "drop") %>%
  mutate(usable = as.numeric(usable)) %>%
  mutate(Sample = factor(Sample, levels=LEVELS)) %>%
  arrange(Sample)

METRICS
```
```{r fig.width=10}

METRICS %>% 
  ggplot(aes(x = Sample, y = usable, fill = Sample)) + geom_bar(stat = 'identity') +
  scale_fill_colorblind() +
  ggtitle("Number of usable reads") +
  theme_bw()

```

```{r truth-reformatting, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
#Clean up TURTH set to match new dataframe (if necessary)
TRUTH = TRUTH %>% 
  mutate(AApos=POS, AltAA=altAA, RefAA=REF)
```

Table Description

  * `chr`: chromosome
  * `POS`: genomic DNA start position
  * `REF`: reference allele
  * `ALT`: alternate allele
  * `AApos`: amino acid position of changes
  * `AltAA`: The new amino acid(s)
  * `AltCodon`: The new codon that leads to this amino acid
  * `EventCount`: number of reads where this change occurred
  * `EventType`: What type of event was observed:
    * `UpstreamNoncoding`: Only upstream (lower positions) noncoding changes
    * `PartialCodingAndUpstreamNoncoding`: A mixture of noncodingregions and incomplete codon (lower positions)
    * `PartialCodingUpStream`:If event occurred in a partial codon upstream (lower positions)
    * `DownstreamNonCoding`: Only downstream stream (higher positions) noncoding changes 
    * `DownstreamNonCodingAndPartialCodingDownstream`:A mixture of noncodingregions and incomplete codon (higher positions)
    * `PartialCodingDownStream`: If event occurred in a partial codon downstream (higher positions)
    * `Synonymous`: No amino acids changed, but DNA changes were present
    * `Missenese`: Single amino acid change
    * `StopGain`: Introduction of a `\*`
    * Reads were excluded if they contained a frameshift, had no changes, too many changes, multiple types of changes, etc.
  * `RefAA`: Reference allele
  * `RefCodon`: reference codon that leads to the reference amino acid
  * `SpliceAI`-fields:
    * `ALLELE`:	Alternate allele
    * `SYMBOL`:	Gene symbol
    * `DS_AG`:	Delta score (acceptor gain)
    * `DS_AL`:	Delta score (acceptor loss)
    * `DS_DG`:	Delta score (donor gain)
    * `DS_DL`:	Delta score (donor loss)
    * `DP_AG`:	Delta position (acceptor gain)
    * `DP_AL`:	Delta position (acceptor loss)
    * `DP_DG`:	Delta position (donor gain)
    * `DP_DL`:	Delta position (donor loss)

```{r warning=FALSE}
Data = NULL
for (x in list.files(data_path, pattern = 'parsed.tsv', full.names = TRUE)) {
  sample_name = strsplit(basename(x), split = '\\.')[[1]][1]
  tmp = read.csv(x, sep = "\t", na.strings = c("None", '.'))
  sample_name = gsub("_hy1|_hyl", "", sample_name)
  tmp$sample_id = sample_name

  #Normalize to usable count
  num_events = METRICS %>%
    filter(Sample == sample_name) %>%
    select(usable) %>%
    as.numeric()
  tmp = tmp %>%
    ungroup() %>%
    group_by(AApos, AltAA) %>%
    mutate(AAEventCount=sum(EventCount, na.rm = TRUE)) %>%
    mutate(AANormalizedEventCount=sum(AAEventCount / num_events)) %>%
    ungroup()
  tmp$NormalizedEventCount = (tmp$EventCount / num_events) 
  tmp$NormalizedEventCount = ifelse(is.na(tmp$NormalizedEventCount), 1e-6,tmp$NormalizedEventCount)
  
  Data = rbind(Data, tmp)
}
Data
```


```{r}
Data = Data %>%
  mutate(AApos = AA_begin + AApos) %>%
  mutate(AA_length = nchar(AltAA)) %>%
  mutate(AA_length = ifelse(is.na(AA_length), 0, AA_length)) %>%
  mutate(SpliceAI_GENE = as.factor(SpliceAI_GENE)) %>%
  mutate(sample_id = as.factor(sample_id)) %>%
  mutate(sample_id = factor(sample_id, levels=LEVELS)) 

data_filename=paste(GENE, FLOWCELL, "Data.tsv", sep=".")
write.table(Data, file = paste(OUTDIR, '/', SAMPLE_PREFIX, data_filename, sep = ""), sep = "\t", row.names = FALSE)
```


# Inter-sample Agreement

## Coding variants
```{r}
Data %>%
  filter(AA_length == 1) %>%
  group_by(AApos, AltCodon, sample_id) %>%
  mutate(N=log(sum(NormalizedEventCount))) %>%
  ungroup() %>%
  select(-contains("SpliceAI")) %>%
  ungroup() %>%
  select(AApos,AltCodon,sample_id, N) %>%
  unique() %>%
  pivot_wider(names_from=sample_id, values_from=N, values_fill=0) %>%
  ungroup() %>%
  select(-AApos, -AltCodon) %>%
  ggpairs(progress = FALSE, title = 'Normalized Counts Agreement') + 
  theme_bw() 
  
```


Evaluate DNA Coverage
```{r fig.width=12}
Data %>%
  group_by(sample_id, EventType) %>%
  ggplot(aes(x = POS, y = NormalizedEventCount, fill = EventType)) +
  geom_bar(stat = 'identity') +
  facet_grid(sample_id ~ ., scales = "free") +
  theme_bw()
```

Nucleotide coverage

```{r fig.height=16, fig.width=12}
Data$EventCount_threshold <- ifelse(Data$EventCount >= 100, "greater_than_100", ifelse((Data$EventCount >= 50 & Data$EventCount < 100), "between50_and_100", ifelse((Data$EventCount >= 20 & Data$EventCount < 50), "between20_and_50", ifelse((Data$EventCount >=5 & Data$EventCount < 20), "between5_and_20",ifelse((Data$EventCount <5), "less_than_5", Data$EventCount)))))
Data$EventCount_threshold <- factor(Data$EventCount_threshold,levels = c("less_than_5", "between5_and_20", "between20_and_50", "between50_and_100", "greater_than_100"))

 Data %>%
  filter(nchar(ALT) == 1) %>%
  filter(ALT != "NA") %>% 
ggplot(aes(x=POS, y=log(EventCount), fill=ALT)) +
    geom_col(aes(fill = ALT), width = 0.7) +
    theme_bw() +
    scale_color_colorblind() +
    facet_grid(sample_id ~ .) 
```

Evaluate Amino acid coverage
```{r fig.height=16, fig.width=12}
Data$EventCount_threshold <- ifelse(Data$EventCount >= 100, "greater_than_100", ifelse((Data$EventCount >= 50 & Data$EventCount < 100), "between50_and_100", ifelse((Data$EventCount >= 20 & Data$EventCount < 50), "between20_and_50", ifelse((Data$EventCount >=5 & Data$EventCount < 20), "between5_and_20",ifelse((Data$EventCount <5), "less_than_5", Data$EventCount)))))


Data$EventCount_threshold <- factor(Data$EventCount_threshold,levels = c("less_than_5", "between5_and_20", "between20_and_50", "between50_and_100", "greater_than_100"))

# No base call (N) and X amino acid numbers 

Data %>%
  filter(AA_length == 1 & AltAA != "X") %>%
  ggplot(aes(x=AApos, y=AltAA, fill=EventType, color=EventType, size=EventCount_threshold, shape=EventType)) +
    geom_point() +
    theme_bw() +
    scale_color_colorblind() +
    facet_grid(sample_id ~ .) 
```


# Summary of Observed - Expected

## At the Single Nucleotide Level 

```{r fig.width=12}
DF_N = 
  Data %>%
   filter(nchar(ALT) == 1) %>%
  filter(ALT != "NA") %>%
  filter(AA_length == 1)%>%
  select(sample_id,AApos, RefAA, AltAA, EventType) %>%
  unique() %>%
  group_by(sample_id, AApos, RefAA, AltAA) %>%
  mutate(Observed = n()) %>%
  ungroup() %>%
  group_by(sample_id, AApos, EventType) %>%
  mutate(ObservedEvents = n()) %>%
  ungroup() %>%
  # select(-Observed,-AltAA) %>%
  # unique() %>%
  pivot_wider(names_from=EventType, values_from=ObservedEvents, values_fill=0) %>%
  mutate(Missing_Missense=19-Missense, Missing_Stop=1-StopGain, Missing_Synonymous=1-Synonymous) %>%
  select(-Missense, -Synonymous, -StopGain,-RefAA) %>%
  mutate(Expected_Missense=19, Expected_Synonymous=1, Expected_Stop=1) %>%
  ungroup() %>%
  unique() %>%
  group_by(sample_id) %>%
  mutate(Total_Expected_Missense=sum(Expected_Missense), Total_Expected_Stop = sum(Expected_Stop), Total_Expected_Synonymous = sum(Expected_Synonymous)) %>%
  mutate(Total_Missing_Missense=sum(Missing_Missense), Total_Missing_Stop = sum(Missing_Stop), Total_Missing_Synonymous = sum(Missing_Synonymous)) %>%
  select(-Missing_Missense, -Missing_Stop, -Missing_Synonymous, -Expected_Missense, -Expected_Synonymous, -Expected_Stop) %>%
  select(-AApos) %>%
  unique() %>%
  pivot_longer(Total_Expected_Missense:Total_Expected_Synonymous, names_to="Expected", values_to="Expected_Counts") %>%
  pivot_longer(Total_Missing_Missense:Total_Missing_Synonymous, names_to="Missing", values_to="Observed_Missing") %>%
  mutate(Expected=gsub("Total_Expected_","",Expected), Missing=gsub("Total_Missing_","", Missing)) %>%
  filter(Expected==Missing) %>%
  select(-Expected) %>%
  mutate(EventType=Missing, Missing=NULL) %>%
  unique() %>% 
  mutate(Proportion_Observed=Observed_Missing/Expected_Counts) %>%
  mutate(sample_id = factor(sample_id, levels = LEVELS))

a = DF_N %>%
  ggplot(aes(x=sample_id, y=Proportion_Observed, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Proportion Missing") +
    xlab("") +
    coord_flip() +
    theme(legend.position = "none")

b = DF_N %>%
  ggplot(aes(x=sample_id, y=Observed_Missing, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    coord_flip() +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Number Missing") +
    xlab("") +
    facet_wrap(EventType ~ ., scales = "free", nrow = 3)

a + b
```
## At the Amino Acid level not combined 

```{r fig.width=12}
DF_N_2 = 
  Data %>%
  filter(AA_length == 1) %>%
  select(sample_id,AApos, RefAA, AltAA, EventType) %>%
  unique() %>%
  group_by(sample_id, AApos, RefAA, AltAA) %>%
  mutate(Observed = n()) %>%
  ungroup() %>%
  group_by(sample_id, AApos, EventType) %>%
  mutate(ObservedEvents = n()) %>%
  ungroup() %>%
  # select(-Observed,-AltAA) %>%
  # unique() %>%
  pivot_wider(names_from=EventType, values_from=ObservedEvents, values_fill=0) %>%
  mutate(Missing_Missense=19-Missense, Missing_Stop=1-StopGain, Missing_Synonymous=1-Synonymous) %>%
  select(-Missense, -Synonymous, -StopGain,-RefAA) %>%
  mutate(Expected_Missense=19, Expected_Synonymous=1, Expected_Stop=1) %>%
  ungroup() %>%
  unique() %>%
  group_by(sample_id) %>%
  mutate(Total_Expected_Missense=sum(Expected_Missense), Total_Expected_Stop = sum(Expected_Stop), Total_Expected_Synonymous = sum(Expected_Synonymous)) %>%
  mutate(Total_Missing_Missense=sum(Missing_Missense), Total_Missing_Stop = sum(Missing_Stop), Total_Missing_Synonymous = sum(Missing_Synonymous)) %>%
  select(-Missing_Missense, -Missing_Stop, -Missing_Synonymous, -Expected_Missense, -Expected_Synonymous, -Expected_Stop) %>%
  select(-AApos) %>%
  unique() %>%
  pivot_longer(Total_Expected_Missense:Total_Expected_Synonymous, names_to="Expected", values_to="Expected_Counts") %>%
  pivot_longer(Total_Missing_Missense:Total_Missing_Synonymous, names_to="Missing", values_to="Observed_Missing") %>%
  mutate(Expected=gsub("Total_Expected_","",Expected), Missing=gsub("Total_Missing_","", Missing)) %>%
  filter(Expected==Missing) %>%
  select(-Expected) %>%
  mutate(EventType=Missing, Missing=NULL) %>%
  unique() %>% 
  mutate(Proportion_Observed=Observed_Missing/Expected_Counts) %>%
  mutate(sample_id = factor(sample_id, levels = LEVELS))
  
a = DF_N_2 %>%
  ggplot(aes(x=sample_id, y=Proportion_Observed, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Proportion Missing") +
    xlab("") +
    coord_flip() +
    theme(legend.position = "none")

b = DF_N_2 %>%
  ggplot(aes(x=sample_id, y=Observed_Missing, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    coord_flip() +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Number Missing") +
    xlab("") +
    facet_wrap(EventType ~ ., scales = "free", nrow = 3)

a + b
```

## At the Amino Acid Level - Combined
```{r fig.width=12}
DF = 
  Data %>%
  filter(AA_length == 1) %>%
  select(sample_id,AApos, RefAA, AltAA, EventType) %>%
  unique() %>%
  group_by(sample_id, AApos, RefAA, AltAA) %>%
  mutate(Observed = n()) %>%
  ungroup() %>%
  group_by(sample_id, AApos, EventType) %>%
  mutate(ObservedEvents = n()) %>%
  ungroup() %>%
  select(-Observed,-AltAA) %>%
  unique() %>%
  pivot_wider(names_from=EventType, values_from=ObservedEvents, values_fill=0) %>%
  mutate(Missing_Missense=19-Missense, Missing_Stop=1-StopGain, Missing_Synonymous=1-Synonymous) %>%
  select(-Missense, -Synonymous, -StopGain,-RefAA) %>%
  mutate(Expected_Missense=19, Expected_Synonymous=1, Expected_Stop=1) %>%
  ungroup() %>%
  unique() %>%
  group_by(sample_id) %>%
  mutate(Total_Expected_Missense=sum(Expected_Missense), Total_Expected_Stop = sum(Expected_Stop), Total_Expected_Synonymous = sum(Expected_Synonymous)) %>%
  mutate(Total_Missing_Missense=sum(Missing_Missense), Total_Missing_Stop = sum(Missing_Stop), Total_Missing_Synonymous = sum(Missing_Synonymous)) %>%
  select(-Missing_Missense, -Missing_Stop, -Missing_Synonymous, -Expected_Missense, -Expected_Synonymous, -Expected_Stop) %>%
  select(-AApos) %>%
  unique() %>%
  pivot_longer(Total_Expected_Missense:Total_Expected_Synonymous, names_to="Expected", values_to="Expected_Counts") %>%
  pivot_longer(Total_Missing_Missense:Total_Missing_Synonymous, names_to="Missing", values_to="Observed_Missing") %>%
  mutate(Expected=gsub("Total_Expected_","",Expected), Missing=gsub("Total_Missing_","", Missing)) %>%
  filter(Expected==Missing) %>%
  select(-Expected) %>%
  mutate(EventType=Missing, Missing=NULL) %>%
  unique() %>% 
  mutate(Proportion_Observed=Observed_Missing/Expected_Counts) %>%
  mutate(sample_id = factor(sample_id, levels = LEVELS))
  
a = DF %>%
  ggplot(aes(x=sample_id, y=Proportion_Observed, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Proportion Missing") +
    xlab("") +
    coord_flip() +
    theme(legend.position = "none")

b = DF %>%
  ggplot(aes(x=sample_id, y=Observed_Missing, fill=EventType)) +
    geom_bar(stat = 'identity', position = 'dodge') +
    coord_flip() +
    theme_bw() +
    scale_fill_colorblind() +
    ylab("Number Missing") +
    xlab("") +
    facet_wrap(EventType ~ ., scales = "free", nrow = 3)

a + b
```


# Results Using Library Control

## Build some functions to calculate ratios
**Note: All Normalized ratios are now at the amino acid level**
**For the Nucleotide level graphs the NormalizedEventCount variable is being used**
```{r}
EVENTS = expand.grid(LEVELS[1], LEVELS[2:length(LEVELS)]) %>% 
  separate(Var2, into = c("Prefix","Day"), remove = FALSE, extra = "merge", sep = Nu_Sep) %>%
  mutate(Event = paste0("mutate(", Day,"_cont_rat = log2(`",Var2,"` / `", Var1,"`))")) %>%
  select(Event) %>%
  unlist() %>%
  paste(collapse = ' %>% ')

# at Amino acid level combined
DL = Data %>% 
  ungroup() %>%
  filter(AA_length == 1) %>%
  select(sample_id, EventType, AApos, RefAA, AltAA, AANormalizedEventCount, -contains('SpliceAI')) %>%
  group_by(sample_id, AApos, RefAA) %>%
  unique() %>%
  ungroup() %>%
  group_by(AApos, RefAA, AltAA) %>%
  pivot_wider(names_from = sample_id, values_from=AANormalizedEventCount)

DLR = eval(parse(text = paste("DL %>%", EVENTS, collapse = '')))

## Single Nuclueotide level 

DL_N = Data %>% 
  ungroup() %>%
  filter(nchar(ALT) == 1) %>%
  filter(ALT != "NA") %>%
  filter(AA_length == 1) %>%
  select(sample_id, EventType, POS, REF, ALT, AApos, RefAA, AltAA, NormalizedEventCount, -contains('SpliceAI')) %>%
  group_by(sample_id, POS, REF) %>%
  unique() %>%
  ungroup() %>%
  group_by(POS, REF, ALT) %>%
  pivot_wider(names_from = sample_id, values_from=NormalizedEventCount)

DLR_N = eval(parse(text = paste("DL_N %>%", EVENTS, collapse = '')))

## ALl Nuclueotide level at single Amino acid change 

DL_N_2 = Data %>% 
  ungroup() %>%
filter(AA_length == 1) %>%
  select(sample_id, EventType, POS, REF, ALT, AApos, RefAA, AltAA, NormalizedEventCount, -contains('SpliceAI')) %>%
  group_by(sample_id, POS, REF) %>%
  unique() %>%
  ungroup() %>%
  group_by(POS, REF, ALT) %>%
  pivot_wider(names_from = sample_id, values_from=NormalizedEventCount)


DLR_N_2 = eval(parse(text = paste("DL_N_2 %>%", EVENTS, collapse = '')))

# Now remove those columns
EVENTS_RM = paste0("select(",paste('-',paste('`',LEVELS,'`', sep = ''), sep = '', collapse = ','),")")


```

## Single Nucleotide Level 

```{r fig.height=6, fig.width=10}
Data_Long_Ratios_N = eval(parse(text = paste("DLR_N %>%", EVENTS_RM,"%>% dplyr::arrange(POS)", collapse = '')))
#vec <- c("pUCSSMmu", "D5", "D8", "D11", "D14", "D20")
N_Count_df <- Data %>% 
    dplyr::select(REF, POS, ALT, EventCount, sample_id) %>%
  filter(!is.na(POS)) %>%
    separate(sample_id, into = c("sample", "Day"), sep = Nu_Sep) %>%
  mutate(variant = paste(REF, POS, ALT, sep="")) %>% 
  dplyr::select(variant, EventCount, Day) %>% 
  group_by(variant, Day) %>%
  mutate(NEventCount=sum(EventCount))%>%
  dplyr::select(variant, Day, NEventCount) %>%
  dplyr::ungroup() %>% unique() %>%
  arrange(match(Day, vec)) %>%
  pivot_wider(names_from = "Day", values_from = "NEventCount", values_fill = list(n = 0)) %>%
  pivot_longer(vec[2]:vec[length(vec)], names_to = "days") %>% 
  mutate(variant_day = paste(variant, days, sep="_"))

colnames(N_Count_df) <- c("variant", "Library_NeventCount", "Days", "NEventCount", "variant_day")


Data_Long_Ratios_N = Data_Long_Ratios_N %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[1]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


new_lib_df_N <- Data_Long_Ratios_N %>%
  mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
  mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
  separate(Ratio, into = c("Day", "metric", "rat"), sep = "_") %>%
  mutate(variant=paste(REF, POS, ALT, sep=""), variant_day=paste(variant, Day, sep="_"))

new_lib_df_N <- merge(new_lib_df_N, N_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_lib_df_N$labels <- paste(paste("variant:",new_lib_df_N$variant.x, sep=""), paste("LibCount:", new_lib_df_N$Library_NeventCount,sep=""), paste("AAcount:",new_lib_df_N$NEventCount, sep=""), sep="|")
new_lib_df_N <- new_lib_df_N[!is.na(new_lib_df_N$SO), ]

q <- ggplot(new_lib_df_N, aes(x=SO, y=Value, fill=SO, label=labels)) +  
  geom_boxplot() + 
  geom_jitter() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(. ~ Day , margins = TRUE)

ggplotly(q)
```


## All Nucleotide Changes - Single Amino acid - Not aggregated 

```{r fig.height=6, fig.width=10}
Data_Long_Ratios_N_2 = eval(parse(text = paste("DLR_N_2 %>%", EVENTS_RM,"%>% dplyr::arrange(POS)", collapse = '')))
#vec <- c("pUCSSMmu", "D5", "D11", "D14", "D20")

cat("The short sample names in the vector vec are", vec, "\n")
cat("=======================================================", "\n")
cat("The conr_rat vector is", Contr_rat, "\n")
cat("=======================================================", "\n")


Data_Long_Ratios_N_2 = Data_Long_Ratios_N_2 %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[1]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


new_lib_df_N_2 <- Data_Long_Ratios_N_2 %>%
  mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
  mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
  separate(Ratio, into = c("Day", "metric", "rat"), sep = "_") %>%
  mutate(variant=paste(REF, POS, ALT, sep=""), variant_day=paste(variant, Day, sep="_"))

new_lib_df_N_2 <- merge(new_lib_df_N_2, N_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_lib_df_N_2$labels <- paste(paste("variant:",new_lib_df_N_2$variant.x, sep=""), paste("LibCount:", new_lib_df_N_2$Library_NeventCount,sep=""), paste("AAcount:",new_lib_df_N_2$NEventCount, sep=""), sep="|")
new_lib_df_N_2 <- new_lib_df_N_2[!is.na(new_lib_df_N_2$SO), ]

q <- ggplot(new_lib_df_N_2, aes(x=SO, y=Value, fill=SO, label=labels)) +  
  geom_boxplot() + 
  geom_jitter() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(. ~ Day , margins = TRUE)

ggplotly(q)
```




## Amino Acid Level - Combined

```{r fig.height=6, fig.width=10}
Data_Long_Ratios = eval(parse(text = paste("DLR %>%", EVENTS_RM,"%>% dplyr::arrange(AApos)", collapse = '')))
#vec <- c("pUCSSMmu", "D5", "D11", "D14", "D20")

cat("The short sample names in the vector vec are", vec, "\n")
cat("=======================================================", "\n")
cat("The conr_rat vector is", Contr_rat, "\n")
cat("=======================================================", "\n")

AA_Count_df <- Data %>% 
  dplyr::select(RefAA, AApos, AltAA, EventCount, sample_id) %>%
  filter(!is.na(AApos)) %>%
  separate(sample_id, into = c("sample", "Day"), sep = Nu_Sep) %>%
  mutate(variant = paste(RefAA, AApos, AltAA, sep="")) %>% 
  dplyr::select(variant, EventCount, Day) %>% 
  group_by(variant, Day) %>%
  mutate(AAEventCount=sum(EventCount))%>%
  dplyr::select(variant, Day, AAEventCount) %>%
  dplyr::ungroup() %>% unique() %>%
  arrange(match(Day, vec)) %>%
  pivot_wider(names_from = "Day", values_from = "AAEventCount", values_fill = list(n = 0)) %>%
  pivot_longer(vec[2]:vec[length(vec)], names_to = "days") %>% 
  mutate(variant_day = paste(variant, days, sep="_"))

colnames(AA_Count_df) <- c("variant", "Library_AAeventCount", "Days", "AAEventCount", "variant_day")


Data_Long_Ratios = Data_Long_Ratios %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[1]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


new_lib_df <- Data_Long_Ratios %>%
  mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
  mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
  separate(Ratio, into = c("Day", "metric"), sep = "_") %>%
  mutate(variant=paste(RefAA, AApos,AltAA, sep=""), variant_day=paste(variant, Day, sep="_"))

new_lib_df <- merge(new_lib_df, AA_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_lib_df$labels <- paste(paste("variant:", new_lib_df$variant.x, sep=""), paste("LibCount:", new_lib_df$Library_AAeventCount,sep=""), paste("AAcount:",new_lib_df$AAEventCount, sep=""), sep="|")


q <- ggplot(new_lib_df, aes(x=SO, y=Value, fill=SO, label=labels)) +  
  geom_boxplot() + 
  geom_jitter() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(. ~ Day , margins = TRUE)

ggplotly(q)
```



# Analyze SpliceAI scores (TBD)
```{r echo=FALSE, warning=FALSE, fig.width=16, fig.height=12}
# Data_Long_Ratios %>%
#   pivot_longer(SpliceAI_DP_AG:SpliceAI_DS_DL, "SpliceAI", values_to="SpliceAI_Score") %>%
#   group_by(Ratio) %>%
#   mutate(PercentRank = percent_rank(Value)) %>%
#   mutate(Group=ifelse(PercentRank<0.1, "Lowest10percent", ifelse(PercentRank>0.9, "Highest10percent", "Other"))) %>%
#   na.omit() %>%
#   ggplot(aes(x=Group, y=log2(SpliceAI_Score))) +
#   geom_boxplot() +
#   facet_grid( SpliceAI ~ Ratio, scales="free") +
#   theme(axis.text.x = element_text(angle = 90, hjust=1)) +
#   stat_compare_means(comparisons = list(c("Lowest10percent", "Highest10percent")) ) +
#   xlab("") +
#   scale_y_continuous(expand = c(0,5))
```

## Identify optimal threshold

## Single Nucleotide level 

```{r}
Target_Ratios_N = Data_Long_Ratios_N %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS_N = unique(na.omit(Target_Ratios_N$Value))

METRICS_TABLE_N = NULL
for (thresh in POSSIBLE_THRESHOLDS_N) {
  predictions = as.factor(ifelse(Target_Ratios_N$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios_N$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE_N = rbind(METRICS_TABLE_N, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}
metrics_filename=paste("metrics_thesholds",GENE,FLOWCELL,"csv", sep=".")
write.csv(METRICS_TABLE_N, file=metrics_filename, sep=',', row.names=F)

OPTIMAL_RESULT_N = METRICS_TABLE_N %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE_N %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT_N$thresh) +
      theme_bw()


OPTIMAL_RESULT_N
```

## All Nucleotide changes - Single Amino Acid - not aggregated 

```{r}
Target_Ratios_N_2 = Data_Long_Ratios_N_2 %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS = unique(na.omit(Target_Ratios_N_2$Value))

METRICS_TABLE_N_2 = NULL
for (thresh in POSSIBLE_THRESHOLDS) {
  predictions = as.factor(ifelse(Target_Ratios_N_2$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios_N_2$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE_N_2 = rbind(METRICS_TABLE_N_2, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}
metrics_filename=paste("metrics_thesholds",GENE,FLOWCELL,"csv", sep=".")
write.csv(METRICS_TABLE_N_2, file=metrics_filename, sep=',', row.names=F)

OPTIMAL_RESULT_N_2 = METRICS_TABLE_N_2 %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE_N_2 %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT_N_2$thresh) +
      theme_bw()


OPTIMAL_RESULT_N_2
```


## Amino Acid Level - aggregated 

```{r}
Target_Ratios = Data_Long_Ratios %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS = unique(na.omit(Target_Ratios$Value))

METRICS_TABLE = NULL
for (thresh in POSSIBLE_THRESHOLDS) {
  predictions = as.factor(ifelse(Target_Ratios$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE = rbind(METRICS_TABLE, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}
metrics_filename=paste("metrics_thesholds",GENE,FLOWCELL,"csv", sep=".")
write.csv(METRICS_TABLE, file=metrics_filename, sep=',', row.names=F)

OPTIMAL_RESULT = METRICS_TABLE %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT$thresh) +
      theme_bw()


OPTIMAL_RESULT
```

## Fixed specificity of 95%

### Single Nucelotide Level 
```{r}
FIXED_SPECIFICITY_N = METRICS_TABLE_N %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
FIXED_SPECIFICITY_N
```



### All Nucleotide changes - Single Amino Acid - not aggregated

```{r}
FIXED_SPECIFICITY_N_2 = METRICS_TABLE_N_2 %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
FIXED_SPECIFICITY_N_2
```

### Amino Acid Level - Aggregated

```{r}
FIXED_SPECIFICITY = METRICS_TABLE %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
FIXED_SPECIFICITY
```

## Compare with Truth Set (If truth provided)

## Single Nucleotide Level 

```{r fig.height=8, fig.width=12, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}

# The truth set (TRUTH) is at the amino acid level and the following cannot be done for the Nucleotide level.

TRUTH_COMPARISON_N = merge(TRUTH, Data_Long_Ratios_N, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON_N = TRUTH_COMPARISON_N %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON_N) > 0) {
  TRUTH_COMPARISON_N %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
    mutate(SO = factor(SO, levels = SO_LEVELS)) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```

## Final Metrics (Optimal) - Single Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N = TRUTH_COMPARISON_N %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N = as.factor(ifelse(Target_Ratios_N$Value < OPTIMAL_RESULT_N$thresh, "Deleterious", "Neutral"))
predictions_N = factor(predictions_N, levels = c("Neutral", "Deleterious"))
truth_N = as.factor(Target_Ratios_N$HDR_Call)
truth_N = factor(truth_N, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N, truth_N, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

### Final Metrics (Fixed Specificity) - Single Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N = TRUTH_COMPARISON_N %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N = as.factor(ifelse(Target_Ratios_N$Value < FIXED_SPECIFICITY_N$thresh, "Deleterious", "Neutral"))
predictions_N = factor(predictions_N, levels = c("Neutral", "Deleterious"))
truth_N = as.factor(Target_Ratios_N$HDR_Call)
truth_N = factor(truth_N, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N, truth_N, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```


### All Nucleotide changes - Single Amino Acid - not aggregated 

```{r fig.height=8, fig.width=12, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}

# The truth set (TRUTH) is at the amino acid level and the following cannot be done for the Nucleotide level.

TRUTH_COMPARISON_N_2 = merge(TRUTH, Data_Long_Ratios_N_2, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON_N_2) > 0) {
  TRUTH_COMPARISON_N_2 %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
    mutate(SO = factor(SO, levels = SO_LEVELS)) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```

### Final Metrics (Optimal) - All Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N_2 = as.factor(ifelse(Target_Ratios_N_2$Value < OPTIMAL_RESULT_N_2$thresh, "Deleterious", "Neutral"))
predictions_N_2 = factor(predictions_N_2, levels = c("Neutral", "Deleterious"))
truth_N_2 = as.factor(Target_Ratios_N_2$HDR_Call)
truth_N_2 = factor(truth_N_2, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N_2, truth_N_2, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

### Final Metrics (Fixed Specificity) - All Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N_2 = as.factor(ifelse(Target_Ratios_N_2$Value < FIXED_SPECIFICITY_N_2$thresh, "Deleterious", "Neutral"))
predictions_N_2 = factor(predictions_N_2, levels = c("Neutral", "Deleterious"))
truth_N_2 = as.factor(Target_Ratios_N_2$HDR_Call)
truth_N_2 = factor(truth_N_2, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N_2, truth_N_2, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

## Amino Acid Level 

```{r fig.height=8, fig.width=12, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}

# The truth set (TRUTH) is at the amino acid level and the following cannot be done for the Nucleotide level.

TRUTH_COMPARISON = merge(TRUTH, Data_Long_Ratios, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON = TRUTH_COMPARISON %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON) > 0) {
  TRUTH_COMPARISON %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat)) %>%
    mutate(SO = factor(SO, levels = SO_LEVELS)) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```

## Final Metrics (Optimal) - Amino Acid Level

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios = TRUTH_COMPARISON %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions = as.factor(ifelse(Target_Ratios$Value < OPTIMAL_RESULT$thresh, "Deleterious", "Neutral"))
predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
truth = as.factor(Target_Ratios$HDR_Call)
truth = factor(truth, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions, truth, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

### Final Metrics (Fixed Specificity) - Amino Acid Level 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios = TRUTH_COMPARISON %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions = as.factor(ifelse(Target_Ratios$Value < FIXED_SPECIFICITY$thresh, "Deleterious", "Neutral"))
predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
truth = as.factor(Target_Ratios$HDR_Call)
truth = factor(truth, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions, truth, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s', err$message)) 
})
```


# Using Day 5 as the Control

```{r}
EVENTS = expand.grid(LEVELS[2], LEVELS[3:length(LEVELS)]) %>% 
  separate(Var2, into = c("Prefix","Day"), remove = FALSE, extra = "merge", sep = Nu_Sep) %>%
  mutate(Event = paste0("mutate(", Day,"_cont_rat = log2(`",Var2,"` / `", Var1,"`))")) %>%
  select(Event) %>%
  unlist() %>%
  paste(collapse = ' %>% ')

DLR = eval(parse(text = paste("DL %>%", EVENTS, collapse = '')))
DLR_N = eval(parse(text = paste("DL_N %>%", EVENTS, collapse = '')))
DLR_N_2 = eval(parse(text = paste("DL_N_2 %>%", EVENTS, collapse = '')))
# Now remove those columns
EVENTS_RM = paste0("select(",paste('-',paste('`',LEVELS,'`', sep = ''), sep = '', collapse = ','),")")
```

## Single Nucleotide Level 

```{r fig.height=6, fig.width=10}
Data_Long_Ratios_N = eval(parse(text = paste("DLR_N %>%", EVENTS_RM,"%>% dplyr::arrange(POS)", collapse = '')))
#vec <- c("pUCSSMmu", "D5", "D8", "D11", "D14", "D20")
N_Count_df <- Data %>% 
    dplyr::select(REF, POS, ALT, EventCount, sample_id) %>%
  filter(!is.na(POS)) %>%
    separate(sample_id, into = c("sample", "Day"), sep = Nu_Sep) %>%
  mutate(variant = paste(REF, POS, ALT, sep="")) %>% 
  dplyr::select(variant, EventCount, Day) %>% 
  group_by(variant, Day) %>%
  mutate(NEventCount=sum(EventCount))%>%
  dplyr::select(variant, Day, NEventCount) %>%
  dplyr::ungroup() %>% unique() %>%
  arrange(match(Day, vec)) %>%
  pivot_wider(names_from = "Day", values_from = "NEventCount", values_fill = list(n = 0)) %>%
  pivot_longer(vec[3]:vec[length(vec)], names_to = "days") %>% 
  mutate(variant_day = paste(variant, days, sep="_"))

colnames(N_Count_df) <- c("variant", "Library_NeventCount", "Days", "NEventCount", "variant_day")


Data_Long_Ratios_N = Data_Long_Ratios_N %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[2]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


new_lib_df_N <- Data_Long_Ratios_N %>%
  mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
  mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
  separate(Ratio, into = c("Day", "metric", "rat"), sep = "_") %>%
  mutate(variant=paste(REF, POS, ALT, sep=""), variant_day=paste(variant, Day, sep="_"))

new_lib_df_N <- merge(new_lib_df_N, N_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_lib_df_N$labels <- paste(paste("variant:",new_lib_df_N$variant.x, sep=""), paste("LibCount:", new_lib_df_N$Library_NeventCount,sep=""), paste("AAcount:",new_lib_df_N$NEventCount, sep=""), sep="|")
new_lib_df_N <- new_lib_df_N[!is.na(new_lib_df_N$SO), ]

q <- ggplot(new_lib_df_N, aes(x=SO, y=Value, fill=SO, label=labels)) +  
  geom_boxplot() + 
  geom_jitter() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(. ~ Day , margins = TRUE)

ggplotly(q)
```


## All Nucleotide Changes - Single Amino acid - Not aggregated 

```{r fig.height=6, fig.width=10}
Data_Long_Ratios_N_2 = eval(parse(text = paste("DLR_N_2 %>%", EVENTS_RM,"%>% dplyr::arrange(POS)", collapse = '')))
#vec <- c("pUCSSMmu", "D5", "D11", "D14", "D20")


Data_Long_Ratios_N_2 = Data_Long_Ratios_N_2 %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[2]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


new_lib_df_N_2 <- Data_Long_Ratios_N_2 %>%
  mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
  mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
  separate(Ratio, into = c("Day", "metric", "rat"), sep = "_") %>%
  mutate(variant=paste(REF, POS, ALT, sep=""), variant_day=paste(variant, Day, sep="_"))

new_lib_df_N_2 <- merge(new_lib_df_N_2, N_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_lib_df_N_2$labels <- paste(paste("variant:",new_lib_df_N_2$variant.x, sep=""), paste("LibCount:", new_lib_df_N_2$Library_NeventCount,sep=""), paste("AAcount:",new_lib_df_N_2$NEventCount, sep=""), sep="|")
new_lib_df_N_2 <- new_lib_df_N_2[!is.na(new_lib_df_N_2$SO), ]

q <- ggplot(new_lib_df_N_2, aes(x=SO, y=Value, fill=SO, label=labels)) +  
  geom_boxplot() + 
  geom_jitter() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(. ~ Day , margins = TRUE)

ggplotly(q)
```

## Amino Acid Level 
```{r fig.height=6, fig.width=12}
Data_Long_Ratios = eval(parse(text = paste("DLR %>%", EVENTS_RM, collapse = '')))

Data_Long_Ratios = Data_Long_Ratios %>%
  pivot_longer(names_to = "Ratio", values_to = "Value", cols = Contr_rat[2]:Contr_rat[length(Contr_rat)]) %>%
  mutate(SO=as.factor(EventType))


AA_Count_df <- Data %>%
  dplyr::select(RefAA, AApos, AltAA, EventCount, sample_id) %>%
  filter(!is.na(AApos)) %>%
  separate(sample_id, into = c("sample", "Day"), sep = Nu_Sep) %>%
  mutate(variant = paste(RefAA, AApos, AltAA, sep="")) %>%
  dplyr::select(variant, EventCount, Day) %>%
  group_by(variant, Day) %>%
  mutate(AAEventCount=sum(EventCount))%>%
  dplyr::select(variant, Day, AAEventCount) %>%
  dplyr::ungroup() %>% unique() %>%
  arrange(match(Day, vec)) %>%
  pivot_wider(names_from = "Day", values_from = "AAEventCount", values_fill = list(n = 0)) %>%
  pivot_longer(vec[3]:vec[length(vec)], names_to = "days") %>%    #modify this lines
  mutate(variant_day = paste(variant, days, sep="_"))

#AA_Count_df
colnames(AA_Count_df) <- c("variant", "Library_AAeventCount", "D5_count", "Days", "AAEventCount", "variant_day")
new_day_df <- Data_Long_Ratios %>% 
	   mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
	   mutate(SO = factor(EventType, levels = SO_LEVELS)) %>%
	   separate(Ratio, into = c("Day", "metric"), sep = "_") %>%
	   mutate(variant=paste(RefAA, AApos,AltAA, sep=""), variant_day=paste(variant, Day, sep="_"))

new_day_df <- merge(new_day_df, AA_Count_df, by.x = "variant_day", by.y="variant_day", all.x = T)
new_day_df$labels <- paste(paste("variant:",new_day_df$variant.x, sep=""), paste("D5Count:", new_day_df$D5_count,sep=""), paste("AAcount:",new_day_df$AAEventCount, sep=""), sep="|")

p <- ggplot(new_day_df, aes(x=SO, y=Value, fill=SO, label=labels)) +
  geom_boxplot() +
  geom_jitter() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + facet_grid(. ~ Day , margins = TRUE)

ggplotly(p)
					  

```

## Identify optimal threshold

## Single Nucleotide level 

```{r}
Target_Ratios_N = Data_Long_Ratios_N %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS_N = unique(na.omit(Target_Ratios_N$Value))

METRICS_TABLE_N = NULL
for (thresh in POSSIBLE_THRESHOLDS_N) {
  predictions = as.factor(ifelse(Target_Ratios_N$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios_N$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE_N = rbind(METRICS_TABLE_N, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}
metrics_filename=paste("metrics_thesholds",GENE,FLOWCELL,"csv", sep=".")
write.csv(METRICS_TABLE_N, file=metrics_filename, sep=',', row.names=F)

OPTIMAL_RESULT_N = METRICS_TABLE_N %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE_N %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT_N$thresh) +
      theme_bw()


OPTIMAL_RESULT_N
```


### Fixed specificity of 95%
```{r}
FIXED_SPECIFICITY_N = METRICS_TABLE_N %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
FIXED_SPECIFICITY_N
```

## All Nucleotide changes - Single Amino Acid - not aggregated 

```{r}
Target_Ratios_N_2 = Data_Long_Ratios_N_2 %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS = unique(na.omit(Target_Ratios_N_2$Value))

METRICS_TABLE_N_2 = NULL
for (thresh in POSSIBLE_THRESHOLDS) {
  predictions = as.factor(ifelse(Target_Ratios_N_2$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios_N_2$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE_N_2 = rbind(METRICS_TABLE_N_2, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}
metrics_filename=paste("metrics_thesholds",GENE,FLOWCELL,"csv", sep=".")
write.csv(METRICS_TABLE_N_2, file=metrics_filename, sep=',', row.names=F)

OPTIMAL_RESULT_N_2 = METRICS_TABLE_N_2 %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE_N_2 %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT_N_2$thresh) +
      theme_bw()


OPTIMAL_RESULT_N_2
```

### Fixed specificity of 95%
```{r}
FIXED_SPECIFICITY_N_2 = METRICS_TABLE_N_2 %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
FIXED_SPECIFICITY_N_2
```


## Amino Acid Level 

```{r}

Target_Ratios = Data_Long_Ratios %>%
  filter(SO %in% c('StopGain', 'Synonymous')) %>%
  filter(Ratio == Optimal_Date) 

POSSIBLE_THRESHOLDS = unique(na.omit(Target_Ratios$Value))

METRICS_TABLE = NULL
for (thresh in POSSIBLE_THRESHOLDS) {
  predictions = as.factor(ifelse(Target_Ratios$Value < thresh, "Deleterious", "Neutral"))
  predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
  truth = as.factor(ifelse(Target_Ratios$SO == 'Synonymous', 'Neutral', 'Deleterious'))
  truth = factor(truth, levels = c("Neutral", "Deleterious"))
  tryCatch(
    {
      cm = confusionMatrix(predictions, truth)
      cm_table = cm$table %>% as.vector()
      TP = cm_table[1]
      FN = cm_table[2]
      FP = cm_table[3]
      TN = cm_table[4]
      METRICS_TABLE = rbind(METRICS_TABLE, data.frame(thresh, t(cm$byClass), TP,FP, FN, TN))
      },
    error = function(e) j = e
  )
}

OPTIMAL_RESULT = METRICS_TABLE %>%
    filter(Balanced.Accuracy == max(Balanced.Accuracy, na.rm = TRUE))

METRICS_TABLE %>%
    select(thresh, Sensitivity, Specificity, Balanced.Accuracy, Precision, Recall) %>%
    pivot_longer(Sensitivity:Recall) %>%
    ggplot(aes(x = thresh, linetype = name, color = name, y = value)) +
      scale_color_gdocs() +
      geom_line() +
      geom_vline(xintercept = OPTIMAL_RESULT$thresh) +
      theme_bw()



OPTIMAL_RESULT
```


## Fixed specificity of 95%
```{r}
FIXED_SPECIFICITY = METRICS_TABLE %>%
  arrange(-Specificity) %>%
  filter(Specificity < 0.95) %>%
  head(1)
```

## Compare with Truth Set (If truth provided)

## At Single Nucleotide Level

```{r fig.height=7, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
TRUTH_COMPARISON_N = merge(TRUTH, Data_Long_Ratios_N, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON_N = TRUTH_COMPARISON_N %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON_N) > 0) {
  TRUTH_COMPARISON_N %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```
## Final Metrics (Optimal) - Single Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N = TRUTH_COMPARISON_N %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N = as.factor(ifelse(Target_Ratios_N$Value < OPTIMAL_RESULT_N$thresh, "Deleterious", "Neutral"))
predictions_N = factor(predictions_N, levels = c("Neutral", "Deleterious"))
truth_N = as.factor(Target_Ratios_N$HDR_Call)
truth_N = factor(truth_N, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N, truth_N, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

### Final Metrics (Fixed Specificity) - Single Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N = TRUTH_COMPARISON_N %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N = as.factor(ifelse(Target_Ratios_N$Value < FIXED_SPECIFICITY_N$thresh, "Deleterious", "Neutral"))
predictions_N = factor(predictions_N, levels = c("Neutral", "Deleterious"))
truth_N = as.factor(Target_Ratios_N$HDR_Call)
truth_N = factor(truth_N, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N, truth_N, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

## All Nucleotide changes - Single Amino Acid - not aggregated 

```{r fig.height=8, fig.width=12, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}

# The truth set (TRUTH) is at the amino acid level and the following cannot be done for the Nucleotide level.

TRUTH_COMPARISON_N_2 = merge(TRUTH, Data_Long_Ratios_N_2, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON_N_2) > 0) {
  TRUTH_COMPARISON_N_2 %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
    mutate(SO = factor(SO, levels = SO_LEVELS)) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```

### Final Metrics (Optimal) - All Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N_2 = as.factor(ifelse(Target_Ratios_N_2$Value < OPTIMAL_RESULT_N_2$thresh, "Deleterious", "Neutral"))
predictions_N_2 = factor(predictions_N_2, levels = c("Neutral", "Deleterious"))
truth_N_2 = as.factor(Target_Ratios_N_2$HDR_Call)
truth_N_2 = factor(truth_N_2, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N_2, truth_N_2, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

### Final Metrics (Fixed Specificity) - All Nucleotide changes 

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios_N_2 = TRUTH_COMPARISON_N_2 %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions_N_2 = as.factor(ifelse(Target_Ratios_N_2$Value < FIXED_SPECIFICITY_N_2$thresh, "Deleterious", "Neutral"))
predictions_N_2 = factor(predictions_N_2, levels = c("Neutral", "Deleterious"))
truth_N_2 = as.factor(Target_Ratios_N_2$HDR_Call)
truth_N_2 = factor(truth_N_2, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions_N_2, truth_N_2, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

## At Amino Acid Level - Aggregated 

```{r fig.height=7, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
TRUTH_COMPARISON = merge(TRUTH, Data_Long_Ratios, by = c('AApos', 'AltAA', 'RefAA')) %>% unique()

TRUTH_COMPARISON = TRUTH_COMPARISON %>%
  filter(!is.na(Value))

if (nrow(TRUTH_COMPARISON) > 0) {
  TRUTH_COMPARISON %>%
    unique() %>%
    mutate(HDR_Call = factor(HDR_Call, levels = c("Neutral", "hypomorphs", "Deleterious"))) %>%
    mutate(Ratio = factor(Ratio, levels = Contr_rat[-1])) %>%
    ggboxplot(x = 'HDR_Call', y = 'Value', color = 'HDR_Call', facet.by = 'Ratio',  add = "jitter") +
    stat_compare_means(comparisons = list(c("Deleterious", "Neutral"))) +
    ylim(-5, 3)
}
```

## Final Metrics (Optimal)

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios = TRUTH_COMPARISON %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions = as.factor(ifelse(Target_Ratios$Value < OPTIMAL_RESULT$thresh, "Deleterious", "Neutral"))
predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
truth = as.factor(Target_Ratios$HDR_Call)
truth = factor(truth, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions, truth, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```

## Final Metrics (Fixed Specificity)

```{r, eval=!is.na(TRUTH), echo=!is.na(TRUTH)}
Target_Ratios = TRUTH_COMPARISON %>%
  filter(SO %in% c('Missense')) %>%
  filter(Ratio == Optimal_Date) 

predictions = as.factor(ifelse(Target_Ratios$Value < FIXED_SPECIFICITY$thresh, "Deleterious", "Neutral"))
predictions = factor(predictions, levels = c("Neutral", "Deleterious"))
truth = as.factor(Target_Ratios$HDR_Call)
truth = factor(truth, levels = c("Neutral", "Deleterious"))
tryCatch({
    cm = confusionMatrix(predictions, truth, positive = "Deleterious")
    print(cm)
},error = function(err){
    (sprintf('Got this error: %s',err$message)) 
})
```


